artiq/language/environment.py

```diff
@@ -126,10 +125,15 @@ class NumberValue(_SimpleArgProcessor):
     :param min: The minimum value of the argument.
     :param max: The maximum value of the argument.
     :param ndecimals: The number of decimals a UI should use.
-    :param force_float: If true, always act as a float even if the value is an integer.
+    :param type: Type of this number. Accepts ``float``, ``int`` or
+                 ``NumberValue.AUTO``. Defaults to ``NumberValue.AUTO``.
     """
+    class AUTO:
+        pass
+    valid_types = [AUTO, float, int]
```

Comment:
Could just be a string instead. This pattern is used commonly in Python...

Reply:
True, that's easier. Give me a minute, I'm still making changes: I'll unmark this PR as draft when I'm done